elki/src/main/java/elki/math/statistics/dependence/MCDEDependenceMeasure.java

```diff
@@ -0,0 +1,249 @@
+import elki.math.statistics.tests.mcde.*;
+
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * Implementation of bivariate Monte Carlo Density Estimation as described in
+ * Edouard Fouché & Klemens Böhm<br>
+ * Monte Carlo Density Estimation<br>
+ * Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)
+ * <p>
+ * This is an abstract class. In order to use MCDE extend it and implement an appropriate statistical test that
+ * returns a p-value and index structure for efficient computation of the statistical test.
+ * <p>
+ * The instantiation of MCDE based on the Mann-Whitney U test is called MWP (as described in the paper).
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+    title = "Monte Carlo Density Estimation", //
+    booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)", url = "https://doi.org/10.1145/3335783.3335795", //
+    bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+public class MCDEDependenceMeasure
+    extends AbstractDependenceMeasure {
+  /**
+   * Monte-Carlo iterations.
+   */
+  protected int m = 50;
+
+  /**
+   * Expected share of instances in slice (independent dimensions).
+   */
+  protected double alpha = 0.5;
+
+  /**
+   * Parameter that specifies the size of the marginal restriction. Note that in the original paper
+   * alpha = beta and as such there is no explicit distinction between the parameters.
+   */
+  protected double beta = 0.5;
+
+  /**
+   * Random generator.
+   */
+  protected RandomFactory rnd;
+
+  /**
+   * Statistical Test returning p-value tailored to MCDE Framework.
+   */
+  protected MCDETest mcdeTest;
+
+  /**
+   * Constructor.
+   */
+  public MCDEDependenceMeasure(int m, double alpha, double beta, RandomFactory rnd, MCDETest mcdeTest) {
+    this.m = m;
+    this.alpha = alpha;
+    this.beta = beta;
+    this.rnd = rnd;
+    this.mcdeTest = mcdeTest;
+  }
+
+  /**
+   * Data Slicing
+   *
+   * @param len         No of data instances
+   * @param nonRefIndex Index (see correctedRank()) computed for the dimension that is not the reference dimension
+   * @return Array of booleans that states which instances are part of the slice
+   */
+  protected boolean[] randomSlice(int len, MCDETest.RankStruct[] nonRefIndex) {
+    final Random random = rnd.getSingleThreadedRandom();
+    boolean slice[] = new boolean[len];
+    Arrays.fill(slice, true);
+
+    // According to the actual formula it should be Math.ceil(Math.pow(this.alpha, 1.0) * len).
+    // The exponent in the multivariate case should be the no. of dimensions - 1 which in the
+    // bivariate case is always simply 1.
+    final int slizeSize = (int) Math.ceil(this.alpha * len);
+    final int start = random.nextInt(len - slizeSize);
+    final int end = start + slizeSize;
+
+    for(int j = 0; j < start; j++) {
+      slice[nonRefIndex[j].index] = false;
```

Comment:
Wouldn't it make sense to iterate start..end, and only set these to true, rather than first filling everything with true, then setting those to false?